frontend/stmt_inferrer.py

```diff
@@ -526,16 +527,33 @@ def _infer_class_def(node, context, solver):
 
                     # handle arguments and return contravariance/covariance
                     for i in range(1, base_args_len):
-                        arg_accessor = getattr(solver.z3_types.type_sort, "func_{}_arg_{}".format(base_args_len, i + 1))
-                        solver.add(solver.z3_types.subtype(arg_accessor(bases_attrs[base][attr]),
-                                                           arg_accessor(class_attrs[attr])),
-                                   fail_message="Arguments contravariance in line {}".format(node.lineno))
-                    return_accessor = getattr(solver.z3_types.type_sort, "func_{}_return".format(base_args_len))
-                    solver.add(solver.z3_types.subtype(return_accessor(class_attrs[attr]),
-                                                       return_accessor(bases_attrs[base][attr])),
-                               fail_message="Return covariance in line {}".format(node.lineno))
-                    solver.optimize.add_soft(return_accessor(class_attrs[attr])
-                                             == return_accessor(bases_attrs[base][attr]))
+                        base_arg_accessor = getattr(solver.z3_types.type_sort, "func_{}_arg_{}"
+                                                    .format(base_args_len, i + 1))
+                        sub_arg_accessor = getattr(solver.z3_types.type_sort, "func_{}_arg_{}"
+                                                   .format(sub_args_len, i + 1))
+                        if attr in inherited_funcs_to_super and inherited_funcs_to_super[attr] == base:
+                            solver.add(base_arg_accessor(bases_attrs[base][attr])
```

Comment:
I'm not sure if this makes a difference, but you could also just assert that the entire function type is the same, right? Instead of doing it for each parameter and for the return type and having (I guess trivially satisfied) constraints about the number of default args?